apis/apiextensions/v1alpha1/ccrd/crd.go

```diff
@@ -42,130 +41,127 @@ const (
 )
 
 const (
-	errNewSpec                 = "cannot generate CustomResourceDefinition from crdSpecTemplate"
+	errGetSpecProps            = "cannot get spec properties from validation schema"
 	errParseValidation         = "cannot parse validation schema"
 	errInvalidClaimNames       = "invalid resource claim names"
 	errMissingClaimNames       = "missing names"
 	errFmtConflictingClaimName = "%q conflicts with composite resource name"
 )
 
-// NOTE(muvaf): We use v1beta1.CustomResourceDefinition for backward
-// compatibility with clusters pre-1.16
-
-// TODO(muvaf): Every field on top level spec could be a DefinitionOption that is
-// reused, although it is known that only two different kinds will be generated.
-
-// An Option configures the supplied CustomResourceDefinition.
-type Option func(*v1beta1.CustomResourceDefinition) error
-
-// New produces a new CustomResourceDefinition.
-func New(o ...Option) (*v1beta1.CustomResourceDefinition, error) {
-	crd := &v1beta1.CustomResourceDefinition{
-		Spec: v1beta1.CustomResourceDefinitionSpec{
-			PreserveUnknownFields: pointer.BoolPtr(false),
-			Subresources: &v1beta1.CustomResourceSubresources{
-				Status: &v1beta1.CustomResourceSubresourceStatus{},
-			},
-			Validation: &v1beta1.CustomResourceValidation{
-				OpenAPIV3Schema: &v1beta1.JSONSchemaProps{
-					Type:       "object",
-					Properties: BaseProps(),
-				},
+// ForCompositeResource derives the CustomResourceDefinition for a composite
+// resource from the supplied CompositeResourceDefinition.
+func ForCompositeResource(xrd *v1alpha1.CompositeResourceDefinition) (*extv1.CustomResourceDefinition, error) {
```

Comment:
There's some arguably drive-by refactoring in this file. I was attempting to update the existing implementations to output v1 CRDs and found that I was struggling to understand what they were doing. I'm hoping that the updated implementation makes a reasonable tradeoff of being a little less DRY in exchange for a little more readability.

Reply:
The new structure looks much simpler to understand, thanks!